File: levelorder.go
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
create a queue with root element
while q is not empty
get size of q, create new array
for loop till size
pop from queue and append value to array
add left, right child of element to q is not nil
append array to result
*/
package main

import "fmt"

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	res := [][]int{}
	if root == nil {
		return res
	}

	q := make([]*TreeNode, 0)
	q = append(q, root)
	for len(q) > 0 {
		size := len(q)
		ele := []int{}
		for i:=0;i<size;i++ {
			val := q[0]
			q = q[1:]
			ele = append(ele, val.Val)
			if val.Left != nil {
				q = append(q, val.Left)
			}
			if val.Right != nil {
				q = append(q, val.Right)
			}
		}
		res = append(res, ele)
	}
	return res
}

func MainLevelOrder() {
	root := &TreeNode{
		Val: 3,
		Left: &TreeNode{
			Val: 9,
		},
		Right: &TreeNode{
			Val: 20,
			Left: &TreeNode{
				Val: 15,
			},
			Right: &TreeNode{
				Val: 7,
			},
		},
	}
	fmt.Println(levelOrder(root)) //expected [[3],[9,20],[15,7]]
}

File: levelOrder.js
// Time Complexity : O(n) where n is no. of tree nodes
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

// BFS - processing parent and its child node together
// BFS uses queue (FIFO)
var levelOrder = function(root) {
    if(!root) return [];
    const queue = [];
    const result = [];
    queue.push(root);
    while(queue.length) {
        const len = queue.length; //No. of nodes to procee at certain level
        let row = [];
        for (let i = 0; i<len; i++) { //For loop to take care of nodes at certain level
            const curr = queue.shift();
            if(curr.left) queue.push(curr.left);
            if(curr.right) queue.push(curr.right);
            row.push(curr.val);
        }
        result.push(row);
    }
    return result;
};